src/kernel/arch/x86/paging.zig

```
@@ -0,0 +1,122 @@
+const ENTRY_WRITABLE = 0x2;
+const ENTRY_USER = 0x4;
+const ENTRY_WRITE_THROUGH = 0x8;
+const ENTRY_CACHE_DISABLED = 0x10;
+const ENTRY_ACCESSED = 0x20;
+const ENTRY_ZERO = 0x40;
+const ENTRY_4MB_PAGES = 0x80;
+const ENTRY_IGNORED = 0x100;
+const ENTRY_AVAILABLE = 0xE00;
+const ENTRY_PAGE_ADDR = 0xFFC00000;
+
+const Directory = packed struct {
+    entries: [ENTRIES_PER_DIRECTORY]DirectoryEntry
+};
+
+///
+/// Map a page directory entry, setting the present, size, writable, write-through and physical address bits.
+/// Clears the user and cache disabled bits.
+///
+/// Arguments:
+///     IN entry: *DirectoryEntry - Unaligned directory entry
+///     IN virt_addr: u32 - The virtual address being mapped
+///     IN phys_addr: u32 - The physical address at which to start mapping
+///     IN allocator: *Allocator - The allocator to use to map any tables needed
+///
+fn map_dir_entry(entry: *align(1) DirectoryEntry, virt_addr: u32, phys_addr: u32, allocator: *std.mem.Allocator) void {
+    entry.* |= ENTRY_PRESENT;
+    entry.* |= ENTRY_WRITABLE;
+    entry.* &= ~u32(ENTRY_USER);
+    entry.* |= ENTRY_WRITE_THROUGH;
+    entry.* &= ~u32(ENTRY_CACHE_DISABLED);
+    entry.* |= ENTRY_4MB_PAGES;
+    entry.* |= ENTRY_PAGE_ADDR & phys_addr;
+}
+
+///
+/// Map a page directory. The addresses passed must be 4KB aligned and be the same distance apart.
+///
+/// Arguments:
+///     IN entry: *Directory - The directory to map
+///     IN phys_start: u32 - The physical address at which to start mapping
+///     IN phys_end: u32 - The physical address at which to stop mapping
+///     IN virt_start: u32 - The virtual address at which to start mapping
+///     IN virt_end: u32 - The virtual address at which to stop mapping
+///     IN allocator: *Allocator - The allocator to use to map any tables needed
+///
+fn map_dir(dir: *Directory, phys_start: usize, phys_end: usize, virt_start: usize, virt_end: usize, allocator: *std.mem.Allocator) void {
+    assert(phys_start <= phys_end);
```

I like the asserts, but if it does trigger, then will panic and may it should just return an error instead, or is panicing good enough

---

Hmm returning an error would be much better as the addresses being malformed isn't fatal, it just means the function can't continue. Will change to an error.